WIP: Skip inlining String indexOf call node with known object receiver

rpshukla opened the pull request:
This PR relies on 2 other PRs to be merged first:

- Transform indexOf call node when receiver is known string: https://github.com/eclipse/openj9/pull/8780
- Add new method in OMR_InlinerPolicy to skip inlining recognized method based on calltarget: https://github.com/eclipse/omr/pull/4940

For String indexOf methods where the receiver is a known object, the call node
can be transformed away by value propagation. Therefore it does not need to be
inlined.

- Override `OMR_InlinerPolicy::shouldSkipInliningRecognizedMethod` in
  `TR_J9InlinerPolicy`

Signed-off-by: Ryan Shukla <ryans@ibm.com>

cathyzhyi commented:
Hey Ryan, thanks for looking into it. This is in reply to your question in https://github.com/eclipse/omr/pull/4940#issuecomment-614266431 but I figured since most of the stuff is openj9 specific I am replying here. After you pointing out the problem of `_calleeSymbol ` always being NULL in `TR_InlinerBase::applyPolicyToTargets`, I realize there are actually 3 issues here:
1. _calleeSymbol being NULL this can be worked around by doing something similar to this code here: https://github.com/eclipse/openj9/blob/d493108cc5cc229002810c55b0ac86df1ca1fbc2/runtime/compiler/env/VMJ9.cpp#L2977-L2980
2. the callNode sometimes can be NULL as well depending on where the `TR_InlinerBase::applyPolicyToTargets` is called. I suspect we might be able to work around this by just doing `ecsArgInfo->get(0)` since the info is either the receiver or the first argument.  And speaking of which, I think we should actually just use ecsArgInfo->get(0); because argInfo only records the real arguments while the `callNode->getFirstArgumentIndex();` starts at 1 for virtual calls because the first child is what we call `vftChild` indicates the receiver class. seeing example below.
```
      n6n      (  0)  treetop                                                                              [0x7fb218780560] bci=[-1,1,20] rc=0 vc=0 vn=- li=- udi=- nc=1
      n5n      (  1)    icalli  java/lang/Object.hashCode()I[#359  virtual Method -48] [flags 0x500 0x0 ]  [0x7fb218780510] bci=[-1,1,20] rc=1 vc=0 vn=- li=- udi=- nc=2
      n4n      (  1)      aloadi  <vft-symbol>[#294  Shadow] [flags 0x18607 0x0 ]                          [0x7fb2187804c0] bci=[-1,1,20] rc=1 vc=0 vn=- li=- udi=- nc=1
      n3n      (  2)        aload  <'this' parm LMain;>[#357  Parm] [flags 0x40000107 0x0 ] (X!=0 sharedMemory )  [0x7fb218780470] bci=[-1,0,20] rc=2 vc=0 vn=- li=- udi=- nc=0 flg=0x4
      n3n      (  2)      ==>aload (X!=0 sharedMemory )
```
3.  Currently`TR_InlinerBase::applyPolicyToTargets` is actually called BEFORE the argInfo gets propagated to the calltargets. I think this is the first time the a target policy involves argInfo but more cases like this might pop up in the future. This last issue makes me think we need a separate item to properly address the issue instead of doing it here. 

In a word, other than fixing the code to use `ecsArgInfo->get(0)` we should probably just call `shouldSkipInliningRecognizedMethod` in `inlineCallTarget2` for now and open a new issue to address it properly. @andrewcraik Hey Andrew, do you have any suggestions on this?

andrewcraik commented:
I think getting this in and opening an item to refactor to make it all a bit nicer sounds reasonable to me.

rpshukla commented:
As per @cathyzhyi's comment, I changed the code to get the recognized method through calleeMethod if calleeSymbol is null like so:
```
   TR::RecognizedMethod rm = calltarget->_calleeSymbol ? calltarget->_calleeSymbol->getRecognizedMethod() : calltarget->_calleeMethod->getRecognizedMethod();
```

I also used `ecsArgInfo->get(0)` instead of `callNode->getFirstArgumentIndex()`.

I updated the omr pr to call `shouldSkipInliningRecognizedMethod` in `inlineCallTarget2` (https://github.com/eclipse/omr/pull/4940)

rpshukla commented:
See comment: https://github.com/eclipse/omr/pull/4940#issuecomment-738145719

genie-openj9 commented:
Can one of the admins verify this patch?
